packages/server/tests/acceptance/ws/subscribe.spec.ts

```diff
@@ -244,15 +248,76 @@ describe('@web-socket Acceptance Tests', async function() {
 
         it('Does not allow more connections than the connection limit', async function() {
             // We already have one connection
+            expect(server._connections).to.equal(1);
+
+            let providers: ethers.providers.WebSocketProvider[] = [];
             for (let i = 1; i < parseInt(process.env.CONNECTION_LIMIT); i++) {
-                await establishConnection();
+                providers.push(await establishConnection());
             }
 
             expect(server._connections).to.equal(parseInt(process.env.CONNECTION_LIMIT));
 
             await expectedErrorAndConnections(server);
 
             await new Promise(resolve => setTimeout(resolve, 1000));
+
+            for (const p of providers) {
+                await p.destroy();
+            }
+
+            await new Promise(resolve => setTimeout(resolve, 1000));
+
+            expect(server._connections).to.equal(1);
+        });
+
+        describe('IP connection limits', async function() {
+            let originalConnectionLimitPerIp;
+
+            before(() => {
+                originalConnectionLimitPerIp = process.env.CONNECTION_LIMIT_PER_IP;
+                process.env.CONNECTION_LIMIT_PER_IP = 3;
+            });
+
+            after(() => {
+                process.env.CONNECTION_LIMIT_PER_IP = originalConnectionLimitPerIp;
+            });
+
+            it('Does not allow more connections from the same IP than the specified limit', async function() {
+                const providers = [];
+                let closeEventHandled = false;
+
+                // Creates the maximum allowed connections
+                for (let i = 1; i < parseInt(process.env.CONNECTION_LIMIT_PER_IP); i++) {
+                    providers.push(await new ethers.providers.WebSocketProvider(WS_RELAY_URL));
+                }
+
+                await new Promise(resolve => setTimeout(resolve, 1000));
+
+                // Repeat the following several times to make sure the internal counters are consistently correct
+                for (let i = 0; i < 3; i++) {
+                    expect(server._connections).to.equal(parseInt(process.env.CONNECTION_LIMIT_PER_IP));
+
+                    // The next connection should be closed by the server
+                    const provider = await new ethers.providers.WebSocketProvider(WS_RELAY_URL);
+
+                    provider._websocket.on('close', (code, message) => {
+                        closeEventHandled = true;
```

Comment:
`closeEventHandled` is never reset to false.
How would a failure when `i > 0` occur?